### PR TITLE
upstream: enable socket_options for multiple local addresses

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: upstream: enable socket_options for multiple local addresses
Additional Description:
Since the ipv4 and ipv6 have different socket option flags, when using multiple addresses, the user has to specify different socket options for the ipv4 address and the ipv6 address.

The endpoint can be ipv4 or ipv6, currently, the user can specify the ipv4 and ipv6 local bind address in the bind config, but there is only a global socket_options that apply to both the ipv4 and ipv6 addresses. Add socket_options for each extra source address.

Risk Level: low
Testing: unittest
Docs Changes: API
Release Notes: new features
Platform Specific Features: n/a

### CODE REVIEW

**Comment by soulxu** [comment]
cc @lizan @mattklein123 this is ready for review

**Review by htuch — COMMENTED** [review]
/lgtm api

File: changelogs/current.yaml [hunk 1]
```diff
@@ -215,6 +215,9 @@ new_features:
 - area: upstream
   change: |
     added a new field :ref:`extra_source_addresses <envoy_v3_api_field_config.core.v3.BindConfig.extra_source_addresses>` to the BindConfig, it enables to specify multiple source addresses, and the source address selection is based on target host's address' version.
+- area: upstream
+  change: |
+    added a new field :ref:`socket_options <envoy_v3_api_field_config.core.v3.ExtraSourceAddress.socket_options>` to the ExtraSourceAddress, it enables to specify the extra source address specified socket options.
```
Comment (mattklein123) [inline changelogs/current.yaml:220]:
```suggestion
    added a new field :ref:`socket_options <envoy_v3_api_field_config.core.v3.ExtraSourceAddress.socket_options>` to the ExtraSourceAddress, allowing specifying discrete socket options for each source address.
```

File: envoy/upstream/upstream.h [hunk 2]
```diff
@@ -38,6 +38,34 @@ class FilterChainManager;
...
+ * A bundle struct for address and socket options.
+ */
+struct UpstreamLocalAddress {
+public:
+  Network::Address::InstanceConstSharedPtr address_;
+  Network::ConnectionSocket::OptionsSharedPtr socket_options_;
+};
+
+/**
+ * A selector for select upstream local address based on the endpoint address.
```
Comment (mattklein123) [inline envoy/upstream/upstream.h:51]:
```suggestion
 * Used to select upstream local address based on the endpoint address.
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this makes sense to me with small nits. Let's merge this after the release is cut. @ravenblackx can you make sure this will work for your use case? Thank you.

/wait

**Review by ravenblackx — APPROVED** [review]
Thanks, looks great! (Sorry for slow review, I changed laptop after starting reviewing!)

**Comment by soulxu** [comment]
@mattklein123 @ravenblackx thanks for the review!

**Comment by soulxu** [comment]
/wait

waiting on 1.24 release

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**